libraries/standard/mqtt/utest/mqtt_lightweight_utest.c

```diff
@@ -235,6 +323,1154 @@ static size_t encodeString( uint8_t * pDestination,
+     * fails  when remaining length is more than packet size. */
+    /* TODO: Since MQTT_SerializeUnsubscribe trusts the remaining length to be
+     * accurate, either change the implementation or remove this test, since it
+     * should fail. */
+    status = MQTT_SerializeUnsubscribe( &subscriptionList,
+                                        subscriptionCount,
+                                        PACKET_ID,
+                                        remainingLength + 4,
+                                        &fixedBuffer );
+    TEST_ASSERT_EQUAL_INT( MQTTNoMemory, status );
+}
+
+/* ========================================================================== */
+
+/**
+ * @brief Tests that MQTT_GetPublishPacketSize works as intended.
+ */
+void test_MQTT_GetPublishPacketSize( void )
+{
+    MQTTPublishInfo_t publishInfo;
+    size_t remainingLength = 0;
+    size_t packetSize;
+    MQTTStatus_t status = MQTTSuccess;
+
+    /* Verify bad paramameters fail. */
+    status = MQTT_GetPublishPacketSize( NULL, &remainingLength, &packetSize );
+    TEST_ASSERT_EQUAL( MQTTBadParameter, status );
+
+    status = MQTT_GetPublishPacketSize( &publishInfo, NULL, &packetSize );
+    TEST_ASSERT_EQUAL( MQTTBadParameter, status );
+
+    status = MQTT_GetPublishPacketSize( &publishInfo, &remainingLength, NULL );
+    TEST_ASSERT_EQUAL( MQTTBadParameter, status );
+
+    /* Empty topic must fail. */
+    memset( ( void * ) &publishInfo, 0x00, sizeof( publishInfo ) );
+    publishInfo.pTopicName = NULL;
+    publishInfo.topicNameLength = TEST_TOPIC_NAME_LENGTH;
+    status = MQTT_GetPublishPacketSize( &publishInfo, &remainingLength, &packetSize );
+    TEST_ASSERT_EQUAL( MQTTBadParameter, status );
+
+    publishInfo.pTopicName = TEST_TOPIC_NAME;
+    publishInfo.topicNameLength = 0;
+    status = MQTT_GetPublishPacketSize( &publishInfo, &remainingLength, &packetSize );
+    TEST_ASSERT_EQUAL( MQTTBadParameter, status );
+
+    /* Packet too large. */
+    memset( ( void * ) &publishInfo, 0x00, sizeof( publishInfo ) );
+    publishInfo.pTopicName = "/test/topic";
+    publishInfo.topicNameLength = sizeof( "/test/topic" );
+    publishInfo.payloadLength = MQTT_MAX_REMAINING_LENGTH;
+    status = MQTT_GetPublishPacketSize( &publishInfo, &remainingLength, &packetSize );
+    TEST_ASSERT_EQUAL( MQTTBadParameter, status );
+
+    publishInfo.payloadLength = MQTT_MAX_REMAINING_LENGTH - publishInfo.topicNameLength - sizeof( uint16_t ) - 1;
+    status = MQTT_GetPublishPacketSize( &publishInfo, &remainingLength, &packetSize );
+    TEST_ASSERT_EQUAL( MQTTBadParameter, status );
+
+    /* Good case succeeds. */
+    publishInfo.pTopicName = "/test/topic";
+    publishInfo.topicNameLength = sizeof( "/test/topic" );
+    publishInfo.pPayload = "";
+    publishInfo.payloadLength = 0;
+    status = MQTT_GetPublishPacketSize( &publishInfo, &remainingLength, &packetSize );
+    TEST_ASSERT_EQUAL( MQTTSuccess, status );
```

Comment:
Don't we have another branch to cover here "if( pPublishInfo->qos > MQTTQoS0 )" for calculating the packet size?

Reply:
That's already handled when testing MQTT_SerializePublish, though I can copy that code over here too if that's desirable

Reply:
Yes. I think that test belong here.